classes.py

```diff
@@ -0,0 +1,186 @@
+            total_cost += staff_member["Total Cost"]
+
+        return total_cost
+            
+    def total_project_cost(self):
+        """Calculates the total monetary cost of the project,
+        including staff cost, hardware cost, and software cost."""
+        total_cost = 0
+        for staff_member in self.project_staff.values():
+            total_cost += staff_member["Total Cost"]
+        
+        for component in self.hardware_components.values():
+            total_cost += component["Cost"]
+        
+        for component in self.software_components.values():
+            total_cost += component["Cost"]
+        
+        return total_cost
+    
+    def cocomo_estimation(self, loc, mode):
+        """Estimates the project cost using the
+        COCOMO model. It accepts two arguments:
+        1. loc = Lines of Code
+        2. mode = Organic, Semi-Detached, or Embedded"""
+
+        constants = {
+            "Organic": {"a": 2.4, "b": 1.05, "c": 2.5, "d": 0.38},
+            "Semi-Detached": {"a": 3.0, "b": 1.12, "c": 2.5, "d": 0.35},
+            "Embedded": {"a": 3.6, "b": 1.20, "c": 2.5, "d": 0.32},
+        }
+
+        # Validate user input to avoid error later
+        if mode in constants:
+            a, b, c, d = constants[mode].values()
+        else:
+            raise ValueError("Invalid mode selected. Please choose 'Organic', 'Semi-Detached', or 'Embedded'.")
+        
+        # Estimate cost using COCOMO formula
+        effort = a * (loc/1000) ** b
+        development_time = c * (effort**d)
+        staff_required = effort/development_time
+
+        # Average cost of GBP 6,040 per staff per month (Based on resources list)
+        average_monthly_staff_cost = 6040
+
+        # Calculate total cost
+        total_cost = effort * average_monthly_staff_cost
+
+        return total_cost
+
+    def write_to_json(self, path):
+        """Writes cost estimation to JSON file."""
+        pass
+
+    def read_json(self, path):
+        """Reads cost estimation from JSON file."""
+        pass
+
+class HardwareComponent:
+    def __init__(self, name, cost, design_weeks, manufacturing_weeks):
+        self.name = name
+        self.cost = cost * 1000  # Cost per quantity thousand
+        self.design_weeks = design_weeks  # Weeks needed to design the component
+        self.manufacturing_weeks = manufacturing_weeks  # Weeks needed to manufacture the component
+
+class SoftwareComponent:
+    def __init__(self, name, cost, design_weeks):
+        self.name = name
+        self.cost = cost
+        self.design_weeks = design_weeks  # Weeks needed to manufacture the component
+
+class StaffMember:
+    def __init__(self, title, cost_per_day, employment_type, weeks_assigned):
+        self.title = title  # Job Title
+        self.cost_per_day = cost_per_day
+        self.employment_type = employment_type  # In-House or Agency
+        self.weeks_assigned = weeks_assigned  # Amount of weeks assigned to the job
```

Comment:
We could get the estimator to automatically assign tasks to members if we did the following things:

1. Assign each staff member a maximum amount of days they can work. For example, we know the project length will be ~12 months or 52 weeks and each and most staff will work 5 days a week, so those staff could have a `workday_cap` of 52 x 5. The advantage of that would be that our programme can do the allocation of tasks for us by selecting staff for tasks from a pool of staff as long as those staff havent met their workday_cap amount of days. 
2. We could record staff member capabilities as strings in a set. So our StaffMember, constructor might look like this instead:   `def __init__(self, title, cost_per_day, employment_type, workday_cap, capabilities): `. The application could use this information as a constraint when assigning tasks.
3. Update the total_design_cost method of the ProjectEstimator class to allocate each week to a suitable member of staff, and sum the result (while tracking how many days each person has available, perhaps in a 'counter' object)

Reply:
That's a great idea, I'll get right to it! Thanks for having a look @bgraham89 !